keps/sig-network/4969-cluster-domain-downward-api/README.md

```diff
@@ -0,0 +1,938 @@
+  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
+- [ ] (R) Graduation criteria is in place
+  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+All Kubernetes Services (and many Pods) have Fully Qualified Domain Names (FQDNs)
+that are constructed using the format `{service}.{namespace}.svc.{clusterDomain}`,
+where `{clusterDomain}` is _typically_ `cluster.local`, but can be reconfigured
+by the cluster administrator.
+
+Currently, there is no way for cluster workloads to query for this domain name,
+leaving them to either use relative domain names or configure it manually.
+
+This KEP proposes adding a new field mapping into the Downward API mechanism, that workloads can use to request the cluster domain name.
+
+<!--
+This section is incredibly important for producing high-quality, user-focused
+documentation such as release notes or a development roadmap. It should be
+possible to collect this information before implementation begins, in order to
+avoid requiring implementors to split their attention between writing release
+notes and implementing the feature itself. KEP editors and SIG Docs
+should help to ensure that the tone and content of the `Summary` section is
+useful for a wide audience.
+
+A good summary is probably at least a paragraph in length.
+
+Both in this section and below, follow the guidelines of the [documentation
+style guide]. In particular, wrap lines to a reasonable length, to make it
+easier for reviewers to cite specific portions, and to minimize diff churn on
+updates.
+
+[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
+-->
+
+## Motivation
+
+Relative domain names are a source of ambiguity: does `get.app` refer to
+the domain registry [get.app](https://get.app/) or the Service `get` in `app`?
+This also becomes problematic for TLS, since there is no way to distinguish
+which of these two cases a certificate applies to.
```

Comment:
I'm not sure what you mean by this. A TLS certificate always has the FQDN in it, doesn't it?

Reply:
Technically no (subjectAltName can be an email address, an IP address, a URI, maybe a few other things). 

But if it's a DNS name then the client validates against the FQDN and not any other domain. Clients that do not are asking for trouble.

Reply:
I just tested this. At least `curl` tests the certificate's names (CN and SANs) against the verbatim hostname in the URL, before expanding the search domain.

So if your FQDN is `foo.bar`, search domain is `bar`, and run `curl https://foo` then the certificate will be validated against `foo`, not `foo.bar`.

Reply:
To `curl`, the FQDN there is `foo.`

Reply:
In the given context, `foo.` does not resolve. `nslookup foo` returns `foo.bar.`. If we consider `foo.` a valid FQDN for the query then we have watered down the term so far that it has become meaningless.

Reply:
For the KEP text, let's try to clarify what we're saying here.